### PR TITLE
Python Console: Jump to previous/next result (#9784)

### LINKED ISSUE #9784: Python Console: Jump to previous/next result
### Is your feature request related to a problem? Please describe.

In the output pane of the Python Console, it can be tedious to inspect a series of lengthy output results.

### Describe the solution you'd like

Provide key bindings to jump to the previous/next output result.

### Additional context

Upcoming PR

### PR DESCRIPTION
**Pull request by JulienCochuyt**
### Link to issue number:

Fixes #9784

### Summary of the issue:

In the output pane of the Python Console, it can be tedious to inspect a series of lengthy output results.

### Description of how this pull request fixes the issue:

Provide key bindings to jump to the previous/next result, select a whole result and clear the output pane.

### Testing performed:

Ported from an add-on in use for a few ~~months~~ [EDIT] years.

### Change log entry:

Python Console: When the focus is on the output pane, alt+up/down jumps to the previous/next output result (add shift for selecting). control+l clears the output pane.

### CODE REVIEW

**Review by michaelDCurran — APPROVED** [review]

**Comment by JulienCochuyt** [comment]
Added the control+L binding to clear the output and rebased onto latest master.

**Comment by JulienCochuyt** [comment]
990d28f: Rebased onto latest master

**Comment by JulienCochuyt** [comment]
Rebased onto latest master and linted.

**Review by LeonarddeR — APPROVED** [review]

**Comment by JulienCochuyt** [comment]
Working on another subject, I think I found a way to improve this PR: Take advantage of the fact the Python Console is a singleton to control it from the NVDA AppModule.
It would allow to use `alt+up/down` instead of `control+up/down` and even allow the gesture to be configured (or aliased to a braille gesture).
It might also allow to add `alt+shift+up/down` for result selection, but I have first to test if the corresponding NVDAObject correctly detects the selection anchor.
What do you think?

**Comment by leonardder** [comment]
@JulienCochuyt 

> What do you think?

I think this makes sense.

**Comment by JulienCochuyt** [comment]
@leonardder: Just pushed the refactoring using the NVDA AppModule.
Default gestures are now `alt+(shift+)up/downArrow` and still `control+l` to clear the output pane.
Handling of the selection anchor works as expected when extending the current selection.

**Comment by dingpengyu** [comment]
Is there any new progress in pr?

**Comment by JulienCochuyt** [comment]
@dingpengyu wrote:
> Is there any new progress in pr?

From my side, everything is done. I'm using this feature on a daily basis through custom patching for about two years.
The PR has been doubly approved, further improved since then, and kept in sync with the master branch...
Please advise if I can do anything more.

**Comment by leonardder** [comment]
well, if @michaelDCurran approved it, I don't think there's anything holding this back for a merge. May be it was missed somehow.

**Comment by feerrenrut** [comment]
It looks there have been a number of changes since Mick last reviewed. I'll request another review so he can look over the new changes.

**Comment by dingpengyu** [comment]
hello
Is there any new progress in current PR?
thanks

**Comment by lukaszgo1** [comment]
@JulienCochuyt Would you be able to resolve conflicts here?

File: source/appModules/nvda.py
Comment (lukaszgo1) [inline source/appModules/nvda.py:164]: `controlTypes.REASON_CARET` -> `controlTypes.OutputReason.CARET`

**Comment by CyrilleB79** [comment]
While at it, please also update the PR's initial descriptions:
* new key binding description
* clear() / ctrl+L in change log

File: source/appModules/nvda.py
Comment (JulienCochuyt) [reply source/appModules/nvda.py]: Done as of ce92a028c7.
Thanks!

**Comment by cary-rowen** [comment]
This is great.

**Comment by JulienCochuyt** [comment]
@CyrilleB79 wrote:
> While at it, please also update the PR's initial descriptions:
> 
> * new key binding description
> * clear() / ctrl+L in change log

Done. Sorry for the lengthy delay.

File: source/gui/__init__.py
Comment (feerrenrut) [inline source/gui/__init__.py:88]: I think this deserves a comment. What happens without this change?

**Review by feerrenrut — CHANGES_REQUESTED** [review]
I've merged in master, and resolved conflicts. I'd like to understand this `isPrevFocusOnNvdaPopup`, but I'll need more ecplanation.

File: source/gui/__init__.py
Comment (JulienCochuyt) [reply source/gui/__init__.py:93]: Without this change, the default gestures bound to the scripts on the NVDA Python Console could not be customized through the Input Gestures dialog. These are the first scripts defined on the NVDA AppModule itself.

Comments added as of fff2ee34e, feel free to amend them as you see fit.

**Review by feerrenrut — APPROVED** [review]
Thanks @JulienCochuyt

**Comment by XLTechie** [comment]
Thanks @JulienCochuyt, I've been looking forward to this one for a long 
time!

**Comment by XLTechie** [comment]
@JulienCochuyt I am seeing some unexpected behavior with this functionality. I can open an issue, but I thought I'd check here first.

1. Control+l doesn't work to clear. I get the bell sound, and no other action.
2. There is no input help for these gestures.
3. What exactly do you mean by previous/next result? I.E. what are results? Running, for example, `help(wx.MessageDialog)`, then pressing alt+up-arrow to move to the previous result, moves instead through different aspects of the class definition. That is not bad, and may be what is desired sometimes, but it is not what I expected with "result".
4. In a new console, try doing something that emits a traceback. Then try, for example, `help(gui.messageBox)`. From the bottom of output from the help call, I would expect alt+up-arrow to take me to the top of output from help, or even the previous command. But instead I am only taken up through the function parameters. Again, that may be useful in some cases, but it is not what I expect from "previous/next **result**". The behavior in number 3 is more potentially useful as a subcommand if it is unintended, but this one--moving just through a parameter list--could also be done with up/down arrow by themselves. Again, I'm not sure if this was what you intended.
5. From the previous example, even when getting to the top of the function docstring, and I am taken to the previous output, the output boundary causes reading of the whole window. I could live with that--it's easy to silence speech--but it wasn't in known issues of the PR.

**Comment by JulienCochuyt** [comment]
@XLTechie,
Thanks for the feedback.
What you describe is more or less the usual behavior, before merging this PR.
What version of NVDA are you using to test this feature?
(There hasn't be an alpha build that includes it yet: alpha-22599,a7f9d47d is 3 commits behind.)

**Comment by XLTechie** [comment]
I apologize, you're right. I was running that alpha instead of my normal 
private master build.
It works much more as expected when I switched to the proper build.

I will delete my comment and this one (after a bit) so as not to pollute 
the comment space.

**Comment by JulienCochuyt** [comment]
@XLTechie,
No problem.
I'm glad it fits your expectations.

**Comment by JulienCochuyt** [comment]
@feerrenrut,
As a follow-up, I realize the new script category "Python Console" assigned to the new scripts on the NVDA app module should probably also be assigned to the global command `activatePythonConsole` (currently "Tools").
My question is "Where should I put the constant definition?"
Currently, it is an attribute of `appModules.nvda` as it is only used there.
My best bet would be to store it instead as an attribute of the `pythonConsole` module and reference it from both the app module and the global commands, turning their current late imports of the module into module-level imports.
Alternatively, it could be stored as an attribute of the `globalCommands` module - as most are - and imported into the app module from there.
What do you prefer?

**Pull request merged**
